File: src/lib/tauri/llm.ts
/**
 * LLM Inference IPC Wrapper
 *
 * Functions for model loading, text generation, and embeddings.
 * Streaming is handled via Tauri events.
 */

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import type { ModelStatus, GenerateRequest, GenerateStats } from './types';

// ============================================================================
// Model Loading
// ============================================================================

export interface LoadModelOptions {
  /** Path to GGUF model file */
  modelPath: string;
  /** Model ID for identification */
  modelId?: string;
  /** Tokenizer model ID (HuggingFace repo) */
  tokenizerId?: string;
  /** Context length (default: 4096) */
  contextLength?: number;
  /** Enable paged attention for long contexts */
  pagedAttention?: boolean;
  /** Chat template (optional) */
  chatTemplate?: string;
}

/**
 * Load a model for inference
 *
 * Uses mistral.rs GgufModelBuilder with automatic device mapping.
 *
 * @param options - Model loading options
 * @returns Model status after loading
 *
 * @example
 * ```ts
 * const status = await loadModel({
 *   modelPath: '/path/to/llama.gguf',
 *   modelId: 'llama-3-8b',
 *   contextLength: 8192,
 * });
 * console.log(`Loaded on ${status.device}`);
 * ```
 */
export async function loadModel(options: LoadModelOptions): Promise<ModelStatus> {
  return invoke<ModelStatus>('load_model', {
    modelPath: options.modelPath,
    modelId: options.modelId,
    tokenizerId: options.tokenizerId,
    contextLength: options.contextLength,
    pagedAttention: options.pagedAttention,
    chatTemplate: options.chatTemplate,
  });
}

/**
 * Get current model status
 *
 * @returns Current model status
 *
 * @example
 * ```ts
 * const status = await getModelStatus();
 * if (status.loaded) {
 *   console.log(`Model: ${status.modelId} on ${status.device}`);
 * }
 * ```
 */
export async function getModelStatus(): Promise<ModelStatus> {
  return invoke<ModelStatus>('get_model_status');
}

/**
 * Unload the current model
 *
 * @example
 * ```ts
 * await unloadModel();
 * ```
 */
export async function unloadModel(): Promise<void> {
  return invoke<void>('unload_model');
}

// ============================================================================
// Text Generation
// ============================================================================

export interface GenerateOptions extends Omit<GenerateRequest, 'prompt'> {
  /** Callback for each generated token */
  onToken?: (token: string) => void;
  /** Callback when generation is complete */
  onDone?: (stats: GenerateStats) => void;
  /** Callback for errors */
  onError?: (error: string) => void;
}

/**
 * Generate text using the loaded model (streaming)
 *
 * Tokens are streamed via the onToken callback.
 *
 * @param prompt - Input prompt
 * @param options - Generation options and callbacks
 * @returns Cleanup function to remove event listeners
 *
 * @example
 * ```ts
 * let fullText = '';
 * const cleanup = await generate('Hello, how are you?', {
 *   maxTokens: 256,
 *   temperature: 0.7,
 *   onToken: (token) => {
 *     fullText += token;
 *     process.stdout.write(token);
 *   },
 *   onDone: (stats) => {
 *     console.log(`\n${stats.tokensPerSecond} tok/s`);
 *   },
 * });
 *
 * // Later: cleanup listeners
 * cleanup();
 * ```
 */
export async function generate(
  prompt: string,
  options: GenerateOptions = {}
): Promise<() => void> {
  const unlisteners: UnlistenFn[] = [];

  // Set up event listeners
  if (options.onToken) {
    const unlisten = await listen<string>('llm:token', (event) => {
      options.onToken?.(event.payload);
    });
    unlisteners.push(unlisten);
  }

  if (options.onDone) {
    const unlisten = await listen<GenerateStats>('llm:done', (event) => {
      options.onDone?.(event.payload);
    });
    unlisteners.push(unlisten);
  }

  if (options.onError) {
    const unlisten = await listen<string>('llm:error', (event) => {
      options.onError?.(event.payload);
    });
    unlisteners.push(unlisten);
  }

  // Start generation
  await invoke<void>('generate', {
    prompt,
    systemPrompt: options.systemPrompt,
    maxTokens: options.maxTokens,
    temperature: options.temperature,
    topP: options.topP,
    stopSequences: options.stopSequences,
  });

  // Return cleanup function
  return () => {
    unlisteners.forEach((unlisten) => unlisten());
  };
}

/**
 * Generate text (non-streaming)
 *
 * Returns the complete generated text.
 *
 * @param prompt - Input prompt
 * @param options - Generation options
 * @returns Generated text
 *
 * @example
 * ```ts
 * const response = await generateSync('What is 2+2?', {
 *   maxTokens: 100,
 * });
 * console.log(response);
 * ```
 */
export async function generateSync(
  prompt: string,
  options: Omit<GenerateOptions, 'onToken' | 'onDone' | 'onError'> = {}
): Promise<string> {
  return invoke<string>('generate_sync', {
    prompt,
    systemPrompt: options.systemPrompt,
    maxTokens: options.maxTokens,
    temperature: options.temperature,
    topP: options.topP,
  });
}

/**
 * Stop ongoing generation
 *
 * @example
 * ```ts
 * await stopGeneration();
 * ```
 */
export async function stopGeneration(): Promise<void> {
  return invoke<void>('stop_generation');
}

// ============================================================================
// Embeddings
// ============================================================================

/**
 * Generate embeddings for text
 *
 * @param texts - Array of texts to embed
 * @returns Array of embedding vectors
 *
 * @example
 * ```ts
 * const embeddings = await embedText(['Hello', 'World']);
 * console.log(`Dimension: ${embeddings[0].length}`);
 * ```
 */
export async function embedText(texts: string[]): Promise<number[][]> {
  return invoke<number[][]>('embed_text', { texts });
}

File: src-tauri/src/commands/llm.rs
//! LLM Commands
//!
//! Tauri commands for LLM inference and embedding using mistral.rs.
//!
//! ## Architecture
//! - Model loading via InferenceEngine (wraps mistral.rs GgufModelBuilder)
//! - Streaming generation via Tauri events
//! - MCP integration via mistralrs_mcp (configured via McpConfigManager)

use std::sync::Arc;
use tauri::{AppHandle, Emitter, State};

use crate::error::{AppError, Result};
use crate::services::{EmbedRequest, GenerateRequest, ModelConfig, ModelStatus};
use crate::state::AppState;

/// Load a model for inference
///
/// Uses mistral.rs GgufModelBuilder with automatic device mapping
#[tauri::command]
pub async fn load_model(
    state: State<'_, Arc<AppState>>,
    model_path: String,
    model_id: Option<String>,
    tokenizer_id: Option<String>,
    context_length: Option<usize>,
    paged_attention: Option<bool>,
    chat_template: Option<String>,
) -> Result<ModelStatus> {
    log::info!("Loading model: {}", model_path);

    let config = ModelConfig {
        model_path: model_path.clone(),
        model_id: model_id.unwrap_or_else(|| {
            std::path::Path::new(&model_path)
                .file_stem()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_else(|| "unknown".to_string())
        }),
        tokenizer_id,
        context_length,
        paged_attention: paged_attention.unwrap_or(false),
        chat_template,
    };

    let mut engine = state.inference_engine.write().await;
    let status = engine.load_model(config).await?;

    log::info!("Model loaded: {} on {}", status.model_id.as_deref().unwrap_or("unknown"), status.device);
    Ok(status)
}

/// Get current model status
#[tauri::command]
pub async fn get_model_status(state: State<'_, Arc<AppState>>) -> Result<ModelStatus> {
    let engine = state.inference_engine.read().await;
    Ok(engine.status().clone())
}

/// Generate text using the loaded model (streaming via events)
///
/// Emits events:
/// - `llm:token`: Each generated token
/// - `llm:done`: Generation complete with stats
/// - `llm:error`: If an error occurs
#[tauri::command]
pub async fn generate(
    app: AppHandle,
    state: State<'_, Arc<AppState>>,
    prompt: String,
    system_prompt: Option<String>,
    max_tokens: Option<usize>,
    temperature: Option<f32>,
    top_p: Option<f32>,
    stop_sequences: Option<Vec<String>>,
) -> Result<()> {
    log::info!(
        "Generating text: {} chars, max_tokens={}",
        prompt.len(),
        max_tokens.unwrap_or(512)
    );

    let engine = state.inference_engine.read().await;

    if !engine.is_loaded() {
        return Err(AppError::Inference("No model loaded".to_string()));
    }

    let request = GenerateRequest {
        prompt,
        system_prompt,
        max_tokens: max_tokens.unwrap_or(512),
        temperature,
        top_p,
        stop_sequences,
        stream: true,
    };

    // Clone app handle for callback
    let app_handle = app.clone();

    // Use real streaming with callback
    let response = engine
        .generate_stream_with_callback(request, move |token| {
            // Emit each token to frontend
            if let Err(e) = app_handle.emit("llm:token", &token) {
                log::error!("Failed to emit token: {}", e);
            }
        })
        .await?;

    app.emit("llm:done", serde_json::json!({
        "prompt_tokens": response.prompt_tokens,
        "completion_tokens": response.completion_tokens,
        "generation_time_ms": response.generation_time_ms,
        "tokens_per_second": response.tokens_per_second,
    }))?;

    log::info!("Generation complete: {} tokens", response.completion_tokens);
    Ok(())
}

/// Generate text (non-streaming)
#[tauri::command]
pub async fn generate_sync(
    state: State<'_, Arc<AppState>>,
    prompt: String,
    system_prompt: Option<String>,
    max_tokens: Option<usize>,
    temperature: Option<f32>,
    top_p: Option<f32>,
) -> Result<String> {
    let engine = state.inference_engine.read().await;

    if !engine.is_loaded() {
        return Err(AppError::Inference("No model loaded".to_string()));
    }

    let request = GenerateRequest {
        prompt,
        system_prompt,
        max_tokens: max_tokens.unwrap_or(512),
        temperature,
        top_p,
        stop_sequences: None,
        stream: false,
    };

    let response = engine.generate(request).await?;
    Ok(response.text)
}

/// Stop ongoing generation
#[tauri::command]
pub async fn stop_generation(state: State<'_, Arc<AppState>>) -> Result<()> {
    log::info!("Stopping generation...");

    let engine = state.inference_engine.read().await;
    engine.stop().await?;

    Ok(())
}

/// Generate embeddings for text
#[tauri::command]
pub async fn embed_text(
    state: State<'_, Arc<AppState>>,
    texts: Vec<String>,
) -> Result<Vec<Vec<f32>>> {
    log::info!("Embedding {} texts", texts.len());

    let engine = state.inference_engine.read().await;

    if !engine.is_loaded() {
        return Err(AppError::Inference("No model loaded".to_string()));
    }

    let request = EmbedRequest { texts };
    let response = engine.embed(request).await?;

    log::info!("Embedding complete: {} vectors of dim {}", response.embeddings.len(), response.dimension);
    Ok(response.embeddings)
}

/// Unload the current model
#[tauri::command]
pub async fn unload_model(state: State<'_, Arc<AppState>>) -> Result<()> {
    log::info!("Unloading model...");

    let mut engine = state.inference_engine.write().await;
    engine.unload()?;

    log::info!("Model unloaded");
    Ok(())
}
